### PR TITLE
[Feature] Add parquet support for node/edge features in chunked data

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
Support node/edge features with parquet format in chunked data when doing data pre-processing. The provided parquet files are assumed to store only one NdArray, and its dim is default to 2 if not explicitly specified in the file. 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: tools/distpartitioning/dataset_utils.py [hunk 1]
```diff
@@ -126,10 +128,13 @@ def get_dataset(input_dir, graph_name, rank, world_size, schema_map):
                     if not os.path.isabs(nfeat_file):
                         nfeat_file = os.path.join(input_dir, nfeat_file)
                     logging.info(f'Loading node feature[{feat_name}] of ntype[{ntype_name}] from {nfeat_file}')
-                    nfeat.append(np.load(nfeat_file))
+                    nfeat.append(
+                        array_readwriter.get_array_parser(
```
Comment (Rhett-Ying) [inline tools/distpartitioning/dataset_utils.py:132]: could we have a wrapper that could help create `fmt_meta` for the caller? such as `utils.read(nfeat_file)`?

File: tests/tools/test_array_readwriter.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import tempfile
```
Comment (Rhett-Ying) [inline tests/tools/test_array_readwriter.py:1]: how about renaming the test file into `test_utils.py` as we may add more methods into `utils` and more testcases into this file?

File: tests/tools/test_dist_part.py [hunk 3]
```diff
@@ -111,12 +113,16 @@ def test_chunk_graph(num_chunks):
                     assert isinstance(int(num2), int)
 
         # check node/edge_data
+        suffix = 'npy' if data_fmt=='numpy' else 'parquet'
```
Comment (Rhett-Ying) [inline tests/tools/test_dist_part.py:116]: `parquet` is the common suffix that indicates it's PARQUET files?

**Review by Rhett-Ying — APPROVED** [review]

**Comment by Rhett-Ying** [comment]
pls resolve the conflicts. your based branch is too out-dated...

File: tests/tools/test_dist_part.py (hunk 3 above)
Comment (peizhou001) [reply tests/tools/test_dist_part.py:116]: It looks a common preference, but some other suffix like ```parq``` is also used. As we use ```pyarrow``` to resolve the file , and its official doc use ```parquet```, let use it to keep consistent.

File: tests/tools/test_array_readwriter.py (hunk 2 above)
Comment (peizhou001) [reply tests/tools/test_array_readwriter.py:1]: Sure.

File: tools/utils/array_readwriter/parquet.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        metadata = pyarrow.parquet.read_metadata(path)
+        metadata = metadata.schema.to_arrow_schema().metadata
+        # As parquet data are tabularized, we assume the dim of ndarray is 2.
+        # If not, it should be explictly specified in the file as metadata.
+        shape = metadata.get(b'shape', None)
+        table = pyarrow.parquet.read_table(path, memory_map=True)
+        logging.info("Done reading from %s" % path)
+        arr = table.to_pandas().to_numpy()
+        shape = tuple(eval(shape.decode())) if shape else arr.shape
+        return arr.reshape(shape)
```
Comment (Rhett-Ying) [inline tools/utils/array_readwriter/parquet.py:26]: if data is stored in shape of `[10, 5, 5]`, what's the shape of return value from `read()`? `[10, 25]`?

Comment (peizhou001) [reply tools/utils/array_readwriter/parquet.py:26]: no, return value should also be ```[10,5,5]```, but it require shape info in the file.

File: tools/utils/array_readwriter/__init__.py [hunk 5]
```diff
@@ -1,2 +1,2 @@
-from . import csv, numpy_array
+from . import csv, numpy_array, parquet
```
Comment (Rhett-Ying) [inline tools/utils/array_readwriter/__init__.py:1]: `array_readwriter` exists both under `//tools/utils/` and `//tools/distpartitioning`. this looks weird.
And in this file, it use `from . import csv, numpy_array` which indicates import csv from current directory. but actually not.

Comment (Rhett-Ying) [reply tools/utils/array_readwriter/__init__.py:1]: forget to remove these 2 files and directory?

File: tools/distpartitioning/utils.py [hunk 6]
```diff
@@ -1,18 +1,19 @@
 import json
 import logging
 import os
+from contextlib import contextmanager
```
Comment (Rhett-Ying) [inline tools/distpartitioning/utils.py:4]: why import this here? Is it used directly in this file?

File: tools/distpartitioning/utils.py [hunk 7]
```diff
@@ -1,18 +1,19 @@
...
 
-import dgl
+import constants
 import numpy as np
 import psutil
 import pyarrow
 import torch
+from numpy.lib.format import open_memmap
```
Comment (Rhett-Ying) [inline tools/distpartitioning/utils.py:11]: why import this here? Is it used directly in this file?

File: tools/utils/array_readwriter/parquet.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
+import logging
```
Comment (Rhett-Ying) [inline tools/utils/array_readwriter/parquet.py:1]: duplicate

**Review by Rhett-Ying — CHANGES_REQUESTED** [review]

File: tools/utils/array_readwriter/__init__.py (hunk 5 above)
Comment (peizhou001) [reply tools/utils/array_readwriter/__init__.py:1]: Nice catch, let me delete the redundant directory.

File: tools/distpartitioning/utils.py (hunk 6 above)
Comment (peizhou001) [reply tools/distpartitioning/utils.py:4]: Deleted

File: tools/distpartitioning/utils.py (hunk 7 above)
Comment (peizhou001) [reply tools/distpartitioning/utils.py:11]: Deleted

File: tools/utils/array_readwriter/parquet.py (hunk 8 above)
Comment (peizhou001) [reply tools/utils/array_readwriter/parquet.py:1]: The repo is deleted.

File: tools/distpartitioning/dataset_utils.py [hunk 9]
```diff
@@ -143,12 +144,13 @@ def get_dataset(input_dir, graph_name, rank, world_size, num_parts, schema_map):
     if((dataset_features is not None) and (len(dataset_features) > 0)):
         for ntype_name, ntype_feature_data in dataset_features.items():
             for feat_name, feat_data in ntype_feature_data.items():
-                assert feat_data[constants.STR_FORMAT][constants.STR_NAME] == constants.STR_NUMPY
-
+                assert feat_data[constants.STR_FORMAT][constants.STR_NAME] \
```
Comment (frozenbugs) [inline tools/distpartitioning/dataset_utils.py:147]: Use () instead of \\ if we can't fit in 1 line.

File: tools/distpartitioning/array_readwriter/parquet.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        metadata = metadata.schema.to_arrow_schema().metadata
+        # As parquet data are tabularized, we assume the dim of ndarray is 2.
+        # If not, it should be explictly specified in the file as metadata.
+        shape = metadata.get(b"shape", None)
+        table = pyarrow.parquet.read_table(path, memory_map=True)
+        logging.info("Done reading from %s" % path)
+        arr = table.to_pandas().to_numpy()
+        if not shape:
+            logging.warning(
+                "Shape info not found in the metadata,"
```
Comment (frozenbugs) [inline tools/distpartitioning/array_readwriter/parquet.py:27]: Make full use of the 80 chars before change to new line.

File: tools/distpartitioning/array_readwriter/parquet.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+                "read the data as a 2 dim array. "
+            )
+        shape = tuple(eval(shape.decode())) if shape else arr.shape
+        return arr.reshape(shape)
+
+    def write(self, path, arr):
```
Comment (frozenbugs) [inline tools/distpartitioning/array_readwriter/parquet.py:33]: arr for array? just use array. Let's avoid unnecessary abbr. which increase confusion.

File: tests/tools/test_utils.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+import pytest
+from distpartitioning import array_readwriter
+
+
+@pytest.mark.parametrize(
+    "shape", [[500], [300, 10], [200, 5, 5], [100, 5, 5, 5]]
+)
+@pytest.mark.parametrize("format", ["numpy", "parquet"])
+def test_array_readwriter(format, shape):
+    original_arr = np.random.rand(*shape)
```
Comment (frozenbugs) [inline tests/tools/test_utils.py:14]: ditto. arr -> array

File: tools/distpartitioning/array_readwriter/parquet.py (hunk 10 above)
Comment (peizhou001) [reply tools/distpartitioning/array_readwriter/parquet.py:27]: fixed

File: tools/distpartitioning/array_readwriter/parquet.py (hunk 11 above)
Comment (peizhou001) [reply tools/distpartitioning/array_readwriter/parquet.py:33]: Got it, use array for fix.

File: tests/tools/test_utils.py (hunk 12 above)
Comment (peizhou001) [reply tests/tools/test_utils.py:14]: fixed

File: tools/distpartitioning/array_readwriter/parquet.py [hunk 13]
```diff
@@ -24,18 +24,18 @@ def read(self, path):
         arr = table.to_pandas().to_numpy()
         if not shape:
             logging.warning(
-                "Shape info not found in the metadata,"
-                "read the data as a 2 dim array. "
+                "Shape information not found in the metadata, read the data as "
+                "a 2 dim array. "
```
Comment (frozenbugs) [inline tools/distpartitioning/array_readwriter/parquet.py:28]: Remove unnecessary space after .

File: tests/tools/test_utils.py [hunk 14]
```diff
@@ -0,0 +1,25 @@
+import os
```
Comment (frozenbugs) [inline tests/tools/test_utils.py:1]: Which file is this unit test for?
For all the unit test of path/to/filename.py, the name should be tests/path/to/test_filename.py

Comment (peizhou001) [reply tests/tools/test_utils.py:1]: Let me change back to this style. I used this before , but rui want to add more tests into this file. But it looks better to separate the test.

File: tools/distpartitioning/array_readwriter/parquet.py (hunk 13 above)
Comment (peizhou001) [reply tools/distpartitioning/array_readwriter/parquet.py:28]: fixed

**Review by frozenbugs — APPROVED** [review]
Approve for the code readability.

File: tests/tools/test_dist_part.py [hunk 15]
```diff
@@ -214,8 +217,9 @@ def read_orig_ids(fname):
 
 
 @pytest.mark.parametrize("num_chunks, num_parts, world_size", [[8, 4, 2], [9, 6, 3], [11, 11, 1], [11, 4, 2], [5, 3, 1]])
-def test_pipeline_basics(num_chunks, num_parts, world_size):
-    _test_pipeline(num_chunks, num_parts, world_size)
+@pytest.mark.parametrize("data_fmt", ['numpy', 'parquet'])
```
Comment (Rhett-Ying) [inline tests/tools/test_dist_part.py:220]: could we add a separate test for data format. or many tests will run.

**Review by Rhett-Ying — APPROVED** [review]
